File: exercises/localexplanations/ex_rnw/code/lime.py
import sys
import os
import random
sys.path.insert(0, ".")

import matplotlib.pyplot as plt
from utils.dataset import Dataset
import numpy as np
import itertools
from sklearn import tree
from math import sqrt

def get_grid(model, dataset, points_per_feature=50):
    """
    Retrieve grid data for plotting a two-dimensional graph with `points_per_feature` for each axis.
    The space is created by the hyperparameters' lower and upper values. Only the first two input
    labels are used.

    Parameters:
        model: Classifier which can call a predict method.
        dataset (utils.Dataset): Dataset with access to configspace and input labels.
        points_per_feature: How many points in each dimension.

    Returns:
        u (np.ndarray): x-axis points with shape (`points_per_feature`,).
        v (np.ndarray): y-axis points with shape (`points_per_feature`,).
        z (np.ndarray): Values with shape (`points_per_feature`, `points_per_feature`).
    """

    hps = dataset.get_configspace().get_hyperparameters_dict()
    labels = dataset.get_input_labels()

    x1 = np.linspace(hps[labels[0]].lower,
                     hps[labels[0]].upper, points_per_feature)
    x2 = np.linspace(hps[labels[1]].lower,
                     hps[labels[1]].upper, points_per_feature)

    X = []
    for x in itertools.product(x1, x2):
        X.append(x)

    X = np.array(X)
    y = model.predict(X)

    # Reshape all x
    u = x1
    v = x2
    z = y.reshape(points_per_feature, points_per_feature).T

    return u, v, z


def plot_grid(u, v, z, labels, title=None, embedded=False):
    """
    Uses the grid data to add a color grid to the plot.

    Parameters:
        u (np.ndarray): x-axis points with shape (N,).
        v (np.ndarray): y-axis points with shape (N,).
        z (np.ndarray): Color values with shape (N, N).
        labels (list): Labels for x and y axis.
        embedded (bool): Whether a new figure should be created or not.

    Returns: 
        plt (matplotlib.pyplot or utils.styled_plot.plt): Plot with applied color grid.
    """

    if not embedded:
        plt.figure()

    plt.xlabel(labels[0])
    plt.ylabel(labels[1])
    plt.title(title)

    plt.pcolormesh(u, v, z, cmap='viridis', shading='auto')
    plt.colorbar()
    plt.grid(alpha=0)

    return plt

def plot_points_in_grid(plt, Z=[], y=[], weights=None, colors={}, x_interest=None, size=8):
    """
    Given a plot, add scatter points from `Z` and `x_interest`.

    Parameters:
        plt (matplotlib.pyplot or utils.styled_plot.plt): Plot with color grid.
        Z (np.ndarray): Points with shape (?, 2) which should be added to the plot.
        y (np.ndarray): Target values with shape (?,), of the points added to the plot, determines the colouring of points.
        weights (np.ndarray): Normalized weights with shape (?,), determine the size of points in the plot.
        colors (dict): Returns the color for an y value.
        x_interest (np.ndarray): Single point with shape (2,) whose prediciton we want to explain. If None (default) no point is added.
        size (int): Default size of the markers/points. Default is 8.
    """

    w = 1

    for y_ in list(set(y)):
        idx = np.where(y == y_)[0]

        color = "black"
        if y_ in colors:
            color = colors[y_]

        if weights is not None:
            w = weights[idx]

        plt.scatter(Z[idx, 0], Z[idx, 1], c=color, s=w*size, label=y_)

    if x_interest is not None:
        plt.scatter([x_interest[0]], [x_interest[1]],
                    c='red', s=size, label="POI")


def sample_points(model, dataset, num_points, seed=0):
    """
    Samples points for the two first features. Uses the bounds from configspace again.

    Parameters:
        model: Classifier which can call a predict method.
        dataset (utils.Dataset): Dataset with access to configspace and input labels.
        num_points (int): How many points should be sampled.
        seed (int): Seed to feed random.

    Returns:
        X (np.ndarray): Data with shape (`num_points`, 2)
        y (np.ndarray): Target values with shape (`num_points`,)
    """

    return None


def weight_points(x_interest, Z, kernel_width=0.2):
    """
    For every z in `Z` returns a weight depending on the distance to `x_interest`.

    Parameters:
        x_interest (np.ndarray): Single point with shape (2,) whose prediction we want to explain.
        Z (np.ndarray): Points with shape (?, 2), data which needs to be weighted.
        kernel_width (float): kernel_width value to calculate distance according to exponential kernel.

    Returns:
        weights (np.ndarray): Normalized weights between 0..1 with shape (?,).
    """

    return None


def fit_explainer_model(Z, y, weights=None, seed=0):
    """
    Fits a decision tree.

    Parameters:
        Z (np.ndarray): Points with shape (?, 2), used to fit surrogate model.
        y (np.ndarray): Target values with shape (?,).
        weights (np.ndarray): Normalized weights with shape (?,).
        seed (int): Seed for the decision tree.

    Returns:
        model (DecisionTreeRegressor): Fitted explainer model.
    """

    return None


if __name__ == "__main__":
    from sklearn.svm import SVC

    dataset = Dataset(
        "wheat_seeds", [1, 5], [7], normalize=True, categorical=True)
    (X_train, y_train), (X_test, y_test) = dataset.get_data()

    model = SVC(gamma = 'auto')
    model.fit(X_train, y_train)
    X = dataset.X
    s = 1

    x_interest = np.array([0.31, 0.37])
    points_per_feature = 50
    n_points = 1000
    labels = dataset.get_input_labels()
    colors = {
        0: "purple",
        1: "green",
        2: "orange",
    }

    print("Run `get_grid` ...")
    u, v, z = get_grid(
        model, dataset, points_per_feature=points_per_feature)

    print("Run `plot_grid` ...")
    plt = plot_grid(u, v, z, labels=labels, title="SVM")
    plt.show()

    print("Run `sample_points` ...")
    Z_sampled, y_sampled = sample_points(model, dataset, n_points)

    print("Run `plot_points_in_grid` ...")
    plt = plot_grid(u, v, z, labels=labels, title="SVM + Sampled Points")
    plot_points_in_grid(plt, Z_sampled, y_sampled, colors=colors)
    plt.show()

    print("Run `weight_points` ...")
    weights = weight_points(x_interest, Z_sampled)

    print("Run `plot_points_in_grid` ...")
    plt = plot_grid(u, v, z, labels=labels,
                    title="SVM + Weighted Sampled Points")
    plot_points_in_grid(plt, Z_sampled, y_sampled, weights, colors, x_interest)
    plt.show()

    print("Run `fit_explainer_model` ...")
    explainer = fit_explainer_model(Z_sampled, y_sampled, weights)

    print("Compare models ...")

    plt.subplot(1, 2, 1)
    plot_grid(u, v, z, labels=labels, title="SVM", embedded=True)
    plot_points_in_grid(plt, Z_sampled, y_sampled, weights, colors, x_interest)

    plt.subplot(1, 2, 2)
    u2, v2, z2 = get_grid(
        explainer, dataset, points_per_feature=points_per_feature)
    plot_grid(u2, v2, z2, labels=labels, title="Decision Tree", embedded=True)
    plot_points_in_grid(plt, Z_sampled, y_sampled, weights, colors, x_interest)

    plt.show()

File: attic/exercises/old_06_local_explanations/code/lime.R
library("docstring")
library("ggplot2")
library("rpart")

get_grid <- function(model, dataset, points_per_feature = 50) {
  #' Retrieve grid data for plotting a two-dimensional graph with `points_per_feature` for each axis.
  #' The space is created by the lower and upper values of the respective feature. Only the first two input
  #' labels are used.
  
  #' @param model: Classifier which can call a predict method.
  #' @param dataset (data.frame): Input dataset (only contains two features).
  #' @param points_per_feature (integer(1)): How many points in each dimension.
  
  #' @return  data.frame with three columns: 
  #'      - equidistant grid of first feature
  #'      - equidistant grid of second feature
  #'      - pred: prediction values for given feature input
  
  range_x1 <- range(dataset[, 1])
  range_x2 <- range(dataset[, 2])
  
  x1 <- seq(range_x1[1], range_x1[2], length.out = points_per_feature)
  x2 <- seq(range_x2[1], range_x2[2], length.out = points_per_feature)
  
  X <- data.frame(expand.grid(x1, x2))
  names(X) <- names(dataset)
  pred <- predict(model, X, type = "class")
  
  data.frame(X, pred = pred)
}

plot_grid <- function(grid) {
  #'  Uses the grid data to add a color grid to the plot.
  #'
  #'  @param grid (data.frame): Grid data for plot.
  #'
  #'  @return ggplot: grid data plotted with coloring displaying the prediction 
  #'  surface.
  
  xnam <- names(grid)[1]
  ynam <- names(grid)[2]
  ggplot(grid, aes_string(x = xnam, y = ynam, fill = "pred")) + 
    geom_tile() +
    guides(z = guide_legend(title = "pred")) +
    theme_bw() +
    theme(legend.position = "right")
}

plot_points_in_grid <- function(plt, df, weights = NULL, x_interest = NULL, size = 4L) {
  #' Given a plot, add scatter points from `df` and `x_interest`.
  #' 
  #' @param plt (ggplot): Plot with color grid.
  #' @param df (data.frame): Points which should be added to the plot.
  #' @param weights (numeric): Normalized weights with elements equal to the number of rows in `df`. 
  #' Weights are used to determine the size of the points in the plot. If NULL, size of all points are equal.
  #' @param x_interest (data.frame): Single point (one row dataset) whose prediction we want to explain. 
  #' If NULL (default) no point is added.
  #' @param size (numeric(1)): Default size of the points. Default 4L.
  #' 
  
  if (!is.null(weights)) {
    w <- weights
  } else {
    w <- 1L
  }
  
  xnam <- names(df)[1]
  ynam <- names(df)[2]
  plt <- plt + geom_point(mapping = aes_string(x = xnam, y = ynam, color = "pred"), size = w*size, data = df, alpha = 2) + 
    scale_colour_hue(l = 40)
  
  if (!is.null(x_interest)) {
    x_interest$pred <- "1"
    plt <- plt + geom_point(mapping = aes_string(x = xnam, y = ynam), x_interest, colour = "red")
  }
  
  plt
}

sample_points <- function(model, dataset, num_points, seed = 0) {
  
  #' Samples points for the two first features and uses the model to 
  #' receive a prediction for these sampled points.
  #' 
  #'  @param model: Classifier which can call a predict method.
  #' 	@param dataset (data.frame): Input dataset (only contains two features).
  #'  @param num_points (int): How many points should be sampled.
  #'  @param seed (int): Seed to feed random.
  #' 
  #' @return dataset (data.frame) of sampled data (feature) plus a column 'pred' with 
  #' their obtained prediction of the `model`.
  
  return(NULL)
}

weight_points = function(x_interest, df, kernel_width = 0.2) {
  #' For every x in `df` returns a weight depending on the exponential kernel distance to `x_interest`.
  #' 
  #' @param x_interest (data.frame): Single point (one row dataset) whose prediction we want to explain.
  #' @param df (data.frame): Data which needs to be weighted (later used for surrogate model).
  #' @param kernel_width (float): Kernel width for exponential kernel.
  #' 
  #' @return weights (numeric): Normalized weights between 0 and 1 for all data points in df.
  
  return(NULL)
}

fit_explainer_model <- function(df, weights = NULL, seed = 0) {
  #' Fits a decision tree to the weighted data
  #' 
  #' @param df (data.frame): Data for surrogate model, must include an outcome variable `pred`.
  #' @param weights (numeric): Normalized weights with number of elements equal to number of rows of `df`.
  #' @param seed (int): Seed for the decision tree.
  #' 
  #' @return model (rpart): Fitted explainer model.
  
  return(NULL)
}

if (FALSE) {
  set.seed(2022L)
  library("e1071") # SVM 
  library("gridExtra") # to plot two ggplots next to each other
  
  dataset <- read.csv(file = "datasets/wheat_seeds.csv")
  dataset$Type <- as.factor(dataset$Type)
  table(dataset$Type)
  
  min_max_norm <- function(x) {
    (x - min(x)) / (max(x) - min(x))
  }
  
  dataset  <- dataset[c("Perimeter", "Asymmetry.Coeff", "Type")]
  dataset$Perimeter <- min_max_norm(dataset$Perimeter)
  dataset$Asymmetry.Coeff <- min_max_norm(dataset$Asymmetry.Coeff)
  
  traindata <- dataset[sample(seq_len(nrow(dataset)), round(0.6*nrow(dataset)), replace = TRUE),]
  
  # Fit a SVM to the data
  mod <- svm(Type ~ ., data = traindata)
  dataset$Type = NULL
  
  # Compute counterfactual for first observation
  x_interest <- data.frame(Perimeter = 0.31, Asymmetry.Coeff = 0.37)
  
  # Parameters for method
  points_per_feature <- 50L
  n_points <- 1000L
  
  print("Run 'get_grid' ...")
  grid <- get_grid(model = mod, dataset = dataset, points_per_feature = points_per_feature)
  
  print("Run `plot_grid` ...")
  plot <- plot_grid(grid)
  plot
  
  print("Run `sample_points` ...")
  samp <- sample_points(model = mod, dataset = dataset, num_points = n_points)
  
  print("Run `plot_points_in_grid` ...")
  plot_points_in_grid(plt = plot, df = samp, size = .5)
  
  print("Run `weight_points` ...")
  w <- weight_points(x_interest = x_interest, df = samp, kernel_width = 0.2)
  
  print("Run `plot_points_in_grid` ...")
  plot_points_in_grid(plt = plot, df = samp, weights = w, x_interest = x_interest)
  
  print("Run `fit_explainer_model` ...")
  explainer <- fit_explainer_model(df = samp, weights = w)
  
  print("Compare models ...")
  plt1 <- plot_points_in_grid(plt = plot, df = samp, x_interest = x_interest, size = .5)
  grid2 <- get_grid(model = explainer, dataset = dataset, points_per_feature = points_per_feature)
  plot2 <- plot_grid(grid2)
  plt2 <- plot_points_in_grid(plt = plot2, df = samp, x_interest = x_interest, size = .5)
  
  plt1 <- plt1 + ggtitle("SVM")
  plt2 <- plt2 + ggtitle("Decision Tree Explainer")
  grid.arrange(plt1, plt2, ncol = 2L)
}
